Components/Queues/AsyncQueueConsumer.php

```diff
@@ -137,26 +128,36 @@ protected function process(EndpointInterface $queueEndpoint, MessageInterface $m
      * Overrides the main callback function to convert the AMQPMessage from the queue into a QueueMessage.
      *
      * {@inheritdoc}
-     *
-     * @return callable
      */
     protected function callback(EndpointInterface $endpoint): callable
     {
-        return function (AMQPMessage $message) use ($endpoint) {
+        return function (AMQPMessage $encodedMessage) use ($endpoint) {
             try {
                 $start = microtime(true);
-                $queueMessage = $this->serializer->deserialize($message->getBody(), SerializableInterface::class, $this->getQueueDriver($endpoint)->getFormat());
+                $message = $this->getSerializer()->decode([
+                    'body' => $encodedMessage->getBody(),
+                    'headers' => [],
+                ]);
+
                 $this->consumptionDuration = (microtime(true) - $start) * 1000;
 
-                $queueMessage->setMessageId($message->getDeliveryTag());
+                $message->setMessageId($encodedMessage->getDeliveryTag());
             } catch (\Exception $exception) {
+                $this->getExceptionHandler()->handle($exception, [
+                    'endpoint' => $endpoint,
+                    'body' => $encodedMessage->getBody(),
+                    'headers' => $encodedMessage->get('application_headers')->getNativeData(),
+                ]);
+                $message = new QueueMessage();
+                $message->setMessageId($encodedMessage->getDeliveryTag());
+                $this->getQueueDriver($endpoint)->ack($message);
+
                 $this->consumptionDuration = (microtime(true) - $start) * 1000;
```

Comment:
What about moving all this functionality to its own separate function, like `finalizeProcessing()` or something more catchy? In the end, we want this chunk to emulate what the parent does, so we are repeating ourselves.

You're also forgetting here to reset the `consumptionDuration` property. (parent does it after dispatching the vent)

Reply:
The `consumptionDuration` is not needed to be reset because I am using an assignation `=`

About the other part could be cool to review all the duplicated coded between AbstractConsumer, AbstractAsyncConsumer, QueueConsumer and AsyncQueueConsumer. Not sure if we will have the time now